### DISCUSSION TITLE
[Bug]: Can't install Xdebug using instructions of the website/docs

### DISCUSSION THREAD

**Discussion by tembra** [beyondcode/herd-community#306] [Q&A]
### Operating system version

macOS Sonoma 14.2.1

### System architecture

ARM64 (M1, M2, etc)

### Herd Version

1.3.2

### PHP Version

PHP 8.2.14

### Bug description

I followed the instructions of website/docs to install through `brew` and `pecl` and added values to `php.ini` but it doesn't work at all. Xdebug is not loaded.

I'm yet using fish shell and I identify that during onboard Herd did not include `HERD_PHP_82_INI_SCAN_DIR` on my environment variables. So I needed to set up manually on `~/.config/fish/config.fish`

```
set -Ux HERD_PHP_82_INI_SCAN_DIR "$HOME/Library/Application Support/Herd/config/php/82/"
```

After solved this problem I run `php -v` and got the messages below:

```bash
Xdebug requires Zend Engine API version 420230831.
The Zend Engine API version 420220829 which is installed, is outdated.

PHP 8.2.14 (cli) (built: Dec 23 2023 00:10:54) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.14, Copyright (c) Zend Technologies
    with Zend OPcache v8.2.14, Copyright (c), by Zend Technologies
```

What happened is that when I installed PHP through `brew`, as recommended by the docs, it installed PHP 8.3. Then, when I installed Xdebug through `pecl`, it installed a Xdebug version for PHP 8.3. However Herd is using PHP 8.2 version which is incompatible with installed Xdebug.

Since PHP 8.2 is the default version of Herd, the docs should instruction the user to install PHP with `brew install php@8.2` and even aware the user of this incompatibility, if it needs Xdebug for other specific version of PHP.

To solve the problem I needed to remove Xdebug, remove PHP and then install PHP 8.2 and install Xdebug:

```bash
pecl uninstall xdebug
brew remove php
brew install php@8.2
pecl install xdebug
```

Then  I edit Herd `php.ini` file to set the new path with

```
zend_extension=/opt/homebrew/Cellar/php@8.2/8.2.14/pecl/20220829/xdebug.so
```

and enabled it as my needs with

```
xdebug.mode=develop,coverage,debug
```

Finally when I run `php -v` I got:

```
PHP 8.2.14 (cli) (built: Dec 23 2023 00:10:54) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.14, Copyright (c) Zend Technologies
    with Xdebug v3.3.1, Copyright (c) 2002-2023, by Derick Rethans
    with Zend OPcache v8.2.14, Copyright (c), by Zend Technologies
```

**Comment by buddhaCode** [comment]
I am running into a similar problem. I want to have multiple PHP versions installed. Each with some extra extensions. How can I install them for different versions of PHP? As far as I understand the current approach, I need to reinstall the extensions when I switch the PHP version.

**Comment by waterrowernohrd** [comment]
Same problem here. I need Xdebug for a PHP 8.2  laravel development and a PHP 7.4 magento development. Now we have Herd PRO! Will the pro version be a solution to run PHP 7.4 with Xdebug and PHP 8.2 with Xdebug?

**Reply by buddhaCode** [reply]
This one is fixed with the new Herd version 1.4. Even in the unpaid basic version. Herd brings precompiled xDebug extensions for every PHP version. You just need to activate it in the config. Have a look at the docs: https://herd.laravel.com/docs/1/advanced-usage/xdebug

**Comment by tembra** [comment]
Closing this discussion since Herd 1.4 comes with built-in xdebug extension already pre-compiled for all availables PHP versions and processors architectures.

If you hit this discussion later it may also be applied for any other extension that you would like to install.

So if you need to manage different versions from any extension I would recommend you:
1. Install the extension using `pecl` using which PHP version do you want
2. Copy the compiled file (`*.so` or `*.dll` after Herd's Windows release) to your own folder
3. Repeat steps 1 and 2 for another versions of the same extension
4. Point the path of the extension version in your own folder inside of `php.ini`

**Reply by aurawindsurfing** [reply]
Hi @tembra 

Would that be the "official" way to handle installation of extensions, please?

I would appreciate if @mpociot or someone from Beyondcode would add complete chapter on this to the documentation. I have tried it and gave up on Heard as did not have time to figure it out - I was trying to install oci with pecl

**Reply by buddhaCode** [reply]
I could reproduce the steps @tembra outlined and installed sqlsrv and pdo_sqlsrv. So @aurawindsurfing , that's the way it works.

Brew can handle multiple installed php versions. If you have multiple versions installed, you need to link the desired version in brew first before calling `pecl`, like so:
- `brew unlink php` the default version, currently 8.3 or a specific version `brew unlink php@8.2`
- `brew link php@7.4 --force --overwrite`

[Andreas Möller](https://localheinz.com/articles/2020/05/05/switching-php-versions-when-using-homebrew/) wrote a nice article about that.

But I would appreciate a proper article on the docs, too.

**Reply by tembra** [reply]
@aurawindsurfing Well I may assume that this is the "official" way (exactly between quotes) since it was the way that Herd docs suggested to us to install Xdebug, before starting to distribute it as a built-in extension.

However as I'm not from the Herd team, I can not affirm that.

@buddhaCode Thanks to point the multiple PHP version. I guess that it would be obvious (as the needed to uninstall the previous extension through pecl). However it's always better to give more than less information, even when you think it's not necessary. 🤝

**Reply by aurawindsurfing** [reply]
@tembra  I will give it another shot then. Just got put off by the process.
